From 8c4ebcb2762164ae86f4efebe68cb75322d958ed Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Sat, 1 Jun 2024 23:06:41 +0300
Subject: [PATCH] add API for destroying metadata encryption key

Ported from GrapheneOS platform_system_vold 2025061900 (fuzzed past LineageOS changes).

Change-Id: I2aa9d739195b8e3e242768944349f66c51e19b77
---
 MetadataCrypt.cpp            | 13 +++++++++++++
 MetadataCrypt.h              |  1 +
 VoldNativeService.cpp        |  7 +++++++
 VoldNativeService.h          |  1 +
 binder/android/os/IVold.aidl |  1 +
 5 files changed, 23 insertions(+)

diff --git a/MetadataCrypt.cpp b/MetadataCrypt.cpp
index b7b8e3a..5e334ff 100644
--- a/MetadataCrypt.cpp
+++ b/MetadataCrypt.cpp
@@ -486,5 +486,18 @@ bool destroy_dsu_metadata_key(const std::string& dsu_slot) {
     return android::vold::DeleteDirContentsAndDir(dsu_metadata_key_dir) == android::OK;
 }
 
+bool destroy_mountpoint_metadata_key(const std::string& path) {
+    auto rec = GetEntryForMountPoint(&fstab_default, path);
+    if (rec == nullptr) {
+        return false;
+    }
+    bool res = android::vold::destroyKey(rec->metadata_key_dir + "/key");
+    auto tmp_path = rec->metadata_key_dir + "/tmp";
+    if (pathExists(tmp_path)) {
+        res &= android::vold::destroyKey(tmp_path);
+    }
+    return res;
+}
+
 }  // namespace vold
 }  // namespace android
diff --git a/MetadataCrypt.h b/MetadataCrypt.h
index 6c46237..16ccf46 100644
--- a/MetadataCrypt.h
+++ b/MetadataCrypt.h
@@ -39,6 +39,7 @@ bool defaultkey_setup_ext_volume(const std::string& label, const std::string& bl
                                  std::string* out_crypto_blkdev);
 
 bool destroy_dsu_metadata_key(const std::string& dsu_slot);
+bool destroy_mountpoint_metadata_key(const std::string& path);
 
 }  // namespace vold
 }  // namespace android
diff --git a/VoldNativeService.cpp b/VoldNativeService.cpp
index 9c08aed..ddb6945 100644
--- a/VoldNativeService.cpp
+++ b/VoldNativeService.cpp
@@ -957,6 +957,13 @@ binder::Status VoldNativeService::destroyDsuMetadataKey(const std::string& dsuSl
     return translateBool(destroy_dsu_metadata_key(dsuSlot));
 }
 
+binder::Status VoldNativeService::destroyMetadataKey(const std::string& mountPointPath) {
+    ENFORCE_SYSTEM_OR_ROOT;
+    ACQUIRE_CRYPT_LOCK;
+
+    return translateBool(destroy_mountpoint_metadata_key(mountPointPath));
+}
+
 binder::Status VoldNativeService::getStorageSize(int64_t* storageSize) {
     ENFORCE_SYSTEM_OR_ROOT;
     return translate(GetStorageSize(storageSize));
diff --git a/VoldNativeService.h b/VoldNativeService.h
index 3bda68e..95dd113 100644
--- a/VoldNativeService.h
+++ b/VoldNativeService.h
@@ -166,6 +166,7 @@ class VoldNativeService : public BinderService<VoldNativeService>, public os::Bn
     binder::Status destroyDsuMetadataKey(const std::string& dsuSlot) override;
 
     binder::Status getStorageSize(int64_t* storageSize) override;
+    binder::Status destroyMetadataKey(const std::string& mountPointPath) override;
 };
 
 }  // namespace vold
diff --git a/binder/android/os/IVold.aidl b/binder/android/os/IVold.aidl
index 188760f..bb40547 100644
--- a/binder/android/os/IVold.aidl
+++ b/binder/android/os/IVold.aidl
@@ -136,6 +136,7 @@ interface IVold {
     void bindMount(@utf8InCpp String sourceDir, @utf8InCpp String targetDir);
 
     void destroyDsuMetadataKey(@utf8InCpp String dsuSlot);
+    void destroyMetadataKey(@utf8InCpp String mountPointPath);
 
     long getStorageSize();
 
-- 
2.43.0

